### PR TITLE
[5476] Improve bulk QTS recommend performance

### PR DESCRIPTION
**Pull request by jhenderson**
### Context
https://trello.com/c/cP6Ojw7t/5476-when-bulk-recommending-308-trainees-it-took-a-long-time-performance-test-bulk-recommendation

### Changes proposed in this pull request
- Speed up the processing time for the action `BulkUpdate::RecommendationsUploadsController#update`
- New class `BulkUpdate::RecommendationsUploads::TraineeLookup` which uses hash maps built on-demand supporting trainee lookup by TRN, HESA ID or Trainee ID
- Change `Reports::TraineeReport` so it only loads `degree` and `course` on demand (prevents hundreds of unnecessary DB calls during the creation of each row)

**The processing time has gone from on average 30 seconds to 10-12 seconds.** 

### Guidance to review
- Go through the bulk recommend process locally using prod data to get an realistic idea of performance
- Run the test on `main` branch and then switch to this PR branch to compare times

### Important business

- [ ] Does this PR introduce any PII fields that need to be overwritten or deleted in db/scripts/sanitise.sql?
- [ ] Does this PR change the database schema? If so, have you updated the config/analytics.yml file and considered whether you need to send 'import_entity' events?
- [ ] Do we need to send any updates to DQT as part of the work in this PR?
- [ ] Does this PR need an ADR?

NB: Please notify the #twd_data_insights team and ask for a review if new fields are being added to analytics.yml

### CODE REVIEW

**Comment by jhenderson** [comment]
[When bulk recommending 308 trainees, it took a long time. Performance test bulk recommendation.](https://trello.com/c/cP6Ojw7t/5476-when-bulk-recommending-308-trainees-it-took-a-long-time-performance-test-bulk-recommendation)

File: app/services/bulk_update/recommendations_uploads/validate_trainee.rb
Comment (tavy88) [inline app/services/bulk_update/recommendations_uploads/validate_trainee.rb:119]: Could we do here:

```ruby
  def trainees_by_provider_trainee_id
       @trainees_by_provider_trainee_id ||= begin
          t = trainee_lookup[row.provider_trainee_id]
          @found_with = :provider_trainee_id if t.present?
          t
      end 
``` ?

This way I think you don't need to disable the that Cop.

Comment (jhenderson) [reply app/services/bulk_update/recommendations_uploads/validate_trainee.rb]: I ended up going further and rewrote the code to find the trainees with some refactoring along the way.

File: app/controllers/trainees/award_recommendations_controller.rb
Comment (tavy88) [inline app/controllers/trainees/award_recommendations_controller.rb:9]: So no guard here anymore? ```FeatureService.enabled?(:integrate_with_dqt)``` ?

**Comment by tavy88** [comment]
> Just reviewed this in Prod Data. It was MUCH faster, but when I go back to the 'bulk recommend' page after recommending 331 trainees... It still shows those same 331 trainees available to be bulk recommended. The records also show as 'TRN Received'. Not sure how long its meant to take, but I would've expected at least some to have gone through from the background job already.

@jhenderson can you confirm this is not related to this PR?

**Review by tavy88 — APPROVED** [review]
LGTM but wonder if it's worth testing this with some real data before merging?

File: app/controllers/trainees/award_recommendations_controller.rb
Comment (jhenderson) [reply app/controllers/trainees/award_recommendations_controller.rb:9]: The job already has the guard clause built in - I should haven mentioned that.

**Comment by jhenderson** [comment]
> LGTM but wonder if it's worth testing this with some real data before merging?

@claire-hughez Has already tested it in `productiondata` but the audit records cannot be verified as we don't run jobs in that environment. I tested it locally and could confirm that the audit records were created in the background.

**Comment by claire-hughez** [comment]
@jhenderson please can we fix that before merging then?

**Pull request merged**
